Mavlink custom messages update (#2999)
PX4/PX4-user_guide: en/middleware/mavlink.md

The shared identity of the message is conveyed by the message id, along with a CRC ("`CRC_EXTRA`") that uniquely identifies the message based on its name and id, and the field names and types.
The receiving end of the communication will discard any packet for which the message id and the `CRC_EXTRA` do not match.
-->

## PX4 and MAVLink

PX4 uses _standard_ MAVLink definitions by default for the greatest compatibility with MAVLink ground stations, libraries, and external components such as MAVLink cameras.
These are included from `development.xml` (by default) in the `main` branch.

:::note
Not all "standard" messages, commands and microservices are _implemented_.
:::

PX4 includes the [mavlink/mavlink](https://github.com/mavlink/mavlink) repo as a submodule under [/src/modules/mavlink](https://github.com/PX4/PX4-Autopilot/tree/main/src/modules/mavlink), and generates the MAVLink 2 C header files at build time.

There are are number of XML dialect files in [/mavlink/messages/1.0/](https://github.com/mavlink/mavlink/blob/master/message_definitions/v1.0/).

> The dialect that is built is defined in the variable `CONFIG_MAVLINK_DIALECT` used in [/src/modules/mavlink/CMakeLists.txt](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/CMakeLists.txt).

Author: The value is set in `CONFIG_MAVLINK_DIALECT` rather than `MAVLINK_DIALECT` as indicated, and appears to be set in `default.px4board` for SITL. This seems odd - is the setting there used for ALL boards?

Reviewer: The other boards use `common` by default: https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/Kconfig#L10

Author: Ah. So you'd change `default "common"` there to your XML file if you wanted something custom?

Reviewer: You could, but need to be aware if a board has set `CONFIG_MAVLINK_DIALECT`, it will override it.

Author: Thanks. The interesting bit here is that common is really the default on real hardware - a bit of a worry since testing is mostly in SITL so if you don't do real hardware testing you could well use definitions that aren't available in the real world.

I made all this a bit more clear in https://github.com/PX4/PX4-user_guide/pull/3027

    </message>
```

:::note
Note that this is a cut-down version of the not-yet-implemented [BATTERY_STATUS_V2](https://mavlink.io/en/messages/development.html#BATTERY_STATUS_V2) message with randomly chosen unused id of `11514`.
Here we've put the message in `development.xml`, which is fine for testing and if the message is intended to eventually be part of the standard message set, but you might also put a [custom message](#custom-mavlink-messages) in its own dialect file.
:::

Build PX4 for SITL and confirm that the associated message is generated in `/build/px4_sitl_default/mavlink/common/mavlink_msg_battery_status_demo.h`.

Because `BatteryStatus` already exists you will not need to do anything to create or build it.

### Define the Streaming Class

First create a file named `BATTERY_STATUS_DEMO.hpp` for your streaming class (named after the message to stream) inside the [/src/modules/mavlink/streams](https://github.com/PX4/PX4-Autopilot/tree/main/src/modules/mavlink/streams) directory.

Add the headers for the MAVLink and uORB messages to the top of the file:

```cpp
#include <uORB/topics/battery_status.h>

> #include <v2.0/development/mavlink.h>

Author: You shouldn't need to import mavlink here right? I.e. headers are imported at PX4 level based on the config variable, and by this point you should get the correct selection?

Reviewer: Yes exactly

    }
    uint16_t get_id()
    {
        return get_id_static();
    }
    unsigned get_size()
    {
        return MAVLINK_MSG_ID_BATTERY_STATUS_DEMO_LEN + MAVLINK_NUM_NON_PAYLOAD_BYTES;
    }

private:
    //Subscription to array of uORB battery status instances
    uORB::SubscriptionMultiArray<battery_status_s, battery_status_s::MAX_INSTANCES> _battery_status_subs{ORB_ID::battery_status};
    // SubscriptionMultiArray subscription is needed because battery has multiple instances.
    // uORB::Subscription is used to subscribe to a single-instance topic

    /* do not allow top copying this class */
    MavlinkStreamBatteryStatusDemo(MavlinkStreamBatteryStatusDemo &);
    MavlinkStreamBatteryStatusDemo& operator = (const MavlinkStreamBatteryStatusDemo &);

protected:
    explicit MavlinkStreamBatteryStatusDemo(Mavlink *mavlink) : MavlinkStream(mavlink)
    {}

	bool send() override
	{
		bool updated = false;

		// Loop through _battery_status_subs (subscription to array of BatteryStatus instances)
		for (auto &battery_sub : _battery_status_subs) {
            // battery_status_s is a struct that can hold the battery object topic

> 			battery_status_s battery_status;

Author: Original version imported this as a struct - ie. `struct battery_status_s battery_status;` or similar. Doesn't seem to be necessary to declare that type.

			// Update battery_status and publish only if the status has changed
			if (battery_sub.update(&battery_status)) {
                // mavlink_battery_status_demo_t is the MAVLink message object
				mavlink_battery_status_demo_t bat_msg{};

				bat_msg.id = battery_status.id - 1;
				bat_msg.battery_remaining = (battery_status.connected) ? roundf(battery_status.remaining * 100.f) : -1;

				// check if temperature valid
				if (battery_status.connected && PX4_ISFINITE(battery_status.temperature)) {
					bat_msg.temperature = battery_status.temperature * 100.f;
				} else {
					bat_msg.temperature = INT16_MAX;
				}

                //Send the message
				mavlink_msg_battery_status_demo_send_struct(_mavlink->get_channel(), &bat_msg);
				updated = true;
			}
		}

		return updated;
	}

};
```

Most streaming classes are very similar (see examples in [/src/modules/mavlink/streams](https://github.com/PX4/PX4-Autopilot/tree/main/src/modules/mavlink/streams)):

> - The streaming class derives from [`MavlinkStream`](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_stream.h).

Author: ```suggestion
- The streaming class derives from [`MavlinkStream`](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_stream.h) and is named using the pattern `MavlinkStream``CamelCaseMessageName>`.
```

> - The streaming class derives from [`MavlinkStream`](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_stream.h) and is named using the pattern `MavlinkStream``CamelCaseMessageName>`.

Author: ```suggestion
- The streaming class derives from [`MavlinkStream`](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_stream.h) and is named using the pattern `MavlinkStream``<CamelCaseMessageName>`.
```

- The `public` definitions are "near-boilerplate", allowing PX4 to get an instance of the class (`new_instance()`), and then to use it to fetch the name, id, and size of the message from the MAVLink headers (`get_name()`, `get_name_static()`, `get_id_static()`, `get_id()`, `get_size()`).
  For your own streaming classes these can just be copied and modified to match the values for your MAVLink message.
- The `private` definitions subscribe to the uORB topics that need to be published.
  Here we also define constructors to prevent the definition being copied.
- The `protected` section is where the important work takes place!
  Here we override the `send()` method, copying values from the subscribed uORB topic(s) into appropriate fields in the MAVLink message, and then send the message.

>   In this case the uORB topic has multiple instances: one for each battery.

Author: Downside of battery is that it support multiple instances so more complicated processing. Upside is that we get to discuss. I wasn't aware for example difference between subscription `update()` and `copy()`. Section below needs to be checked carefully.

In order to avoid clogging communications links with messages that aren't needed you may not stream all messages by default, even at low rate.

If you needed, a GCS or other MAVLink API can request that particular messages are streamed at a particular rate using [MAV_CMD_SET_MESSAGE_INTERVAL](https://mavlink.io/en/messages/common.html#MAV_CMD_SET_MESSAGE_INTERVAL).
A particular message can be requested just once using [MAV_CMD_REQUEST_MESSAGE](https://mavlink.io/en/messages/common.html#MAV_CMD_REQUEST_MESSAGE).

## Receiving MAVLink Messages

This section explains how to receive a message over MAVLink and publish it to uORB.

It assumes that we are receiving the `BATTERY_STATUS_DEMO` message and we want to update the (existing) [BatteryStatus uORB message](../msg_docs/BatteryStatus.md) with the contained information.
This is the kind of implementation that you would provide to support a MAVLink battery integration with PX4.

Add the headers for the incoming MAVLink message and the uORB topic to publish to in [mavlink_receiver.h](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_receiver.h#L77):

```cpp
#include <uORB/topics/battery_status.h>

> #include <v2.0/development/mavlink.h>

Author: I presume that we don't need the mavlink import even for custom messages? Presumably this is done automatically based on config value?

```

Add a function signature for a function that handles the incoming MAVLink message in the `MavlinkReceiver` class in
[mavlink_receiver.h](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_receiver.h#L126)

```cpp
void handle_message_battery_status_demo(mavlink_message_t *msg);
```

Normally you would add a uORB publisher for the uORB topic to publish in the `MavlinkReceiver` class in
[mavlink_receiver.h](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_receiver.h#L296).
In this case the [BatteryStatus](../msg_docs/BatteryStatus.md) uORB topic already exists:

```cpp
uORB::Publication<battery_status_s> _battery_pub{ORB_ID(battery_status)};
```

Implement the `handle_message_battery_status_demo` function in [mavlink_receiver.cpp](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_receiver.cpp).

```cpp
void
MavlinkReceiver::handle_message_battery_status_demo(mavlink_message_t *msg)
{

> 	if ((msg->sysid != mavlink_system.sysid) || (msg->compid == mavlink_system.compid)) {

Author: So what instance are we writing to? It looks like maybe "the default instance"? What if there are two mavlink batteries and one "internal" one? How do we sort that out? I don't plan to say too much but we should comment on what is happening.

Reviewer: `uORB::Publication<battery_status_s>` means to use single-instance. For multiple it really depends on what you want to achieve. The simplest (and automatic) is to use an array of `uORB::PublicationMulti`.

Author: @bkueng I don't think that quite answers my question. For the incoming battery status message (the current/real one) we have this single publication, so what we have here is a single topic.

```cpp
uORB::Publication<battery_status_s>	_battery_pub{ORB_ID(battery_status)};
```

But when we're streaming we have multiple instances we iterate through and publish a MAVLink message for each.

My inference is that there are multiple battery instance definitions possible in PX4 but we only write into to one of them - if that is correct (?) then which one - the first item, which I expect is the battery with mavlink instance id `1`?

Further doesn't this imply that while we can emit MAVLink messages for any number of connected batteries, we can only ever have one "incoming" battery connected to PX4 via MAVLink?

Lastly, given your answer, I think that this means that if we wanted to have multiple batteries "incoming" and connected via MAVLink we'd need to use a multiple publication, and take the instance number to work out which one to write to. Right?

I'm not going to block on this but if you answer I'll put in an update on my return.

----

As an aside, I think we/I need to go through uORB docs again and document this both as a guide and as a reference - explaining all these methods for subscription. The current docs are ... insufficient.

Reviewer: > My inference is that there are multiple battery instance definitions possible in PX4 but we only write into to one of them - if that is correct (?) then which one - the first item, which I expect is the battery with mavlink instance id 1?

Yes in that case it would publish to the first uorb instance. The battery case is a bit complex for that reason. The message also has an `id` field that's used for mavlink.

> Lastly, given your answer, I think that this means that if we wanted to have multiple batteries "incoming" and connected via MAVLink we'd need to use a multiple publication, and take the instance number to work out which one to write to. Right?

Yes that's where `PublicationMulti` comes in.

Author: Added explanation in https://github.com/PX4/PX4-user_guide/pull/3027

>             _msg_ca_trajectory.seq_id = _ca_trajectory.seq_id;

Author: This line and above define the mapping from uorb to MAVLInk message CA_TRAJECTORY. I think we might help people by providing the definitions of both of those in the doc for people to copy. They could then run the tutorial end to end and know it makes sense. 

For the uorb topic, would this make sense based on the above values? Advice on coefficients?

```
uint64 timestamp				# time since system start (microseconds)
uint64 time_start_usec			# time for trajectory start (microseconds)
uint64 time_stop_usec			# time for trajectory stop (microseconds)
float32 coefficients				# NOT SURE ABOUT THIS - it matches the type, but is singular, so does not match name
uint32 seq_id					# sequence id
```

What do you think of the idea, and if so, is this message OK or we could just pretend that BATTERY_STATUS say is a new message?

@julianoes Opinions?

Reviewer: I think using an existing one would be a bit simpler. Then we can also directly point to the messages.

Author: Thank you!

Add the line below to the part of the file where all the other streams are included:

```cpp
#include "streams/CA_TRAJECTORY.hpp"
```

Finally append the stream class to the `streams_list` at the bottom of
[mavlink_messages.cpp](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_messages.cpp)

```C
StreamListItem *streams_list[] = {
...
#if defined(CA_TRAJECTORY_HPP)
    create_stream_list_item<MavlinkStreamCaTrajectory>(),
#endif // CA_TRAJECTORY_HPP
...
}
```

The class is now available for streaming, but won't be streamed by default.
We cover that in the next sections.

### Streaming by Default

The easiest way to stream your messages by default (as part of a build) is to add them to [mavlink_main.cpp](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/mavlink/mavlink_main.cpp) in the appropriate message group.

If you search in the file you'll find groups of messages defined in a switch statement:

- `MAVLINK_MODE_NORMAL`: Streamed to a GCS.
- `MAVLINK_MODE_ONBOARD`: Streamed to a companion computer on a fast link, such as Ethernet

> - `MAVLINK_MODE_ONBOARD_LOW_BANDWIDTH`: Streamed to a companion computer on slower link

Reviewer: We use this on skynode, not because of slower link but to reduce traffic afterwards (routing to the GCS)

Author: Cheers!

```suggestion
- `MAVLINK_MODE_ONBOARD_LOW_BANDWIDTH`: Streamed to a companion computer for re-routing to a reduced-traffic link, such as a GCS.
```

- `MAVLINK_MODE_GIMBAL`: Streamed to a gimbal
- `MAVLINK_MODE_EXTVISION`: Streamed to an external vision system
- `MAVLINK_MODE_EXTVISIONMIN`: Streamed to an external vision system on a slower link
- `MAVLINK_MODE_OSD`: Streamed to an OSD, such as an FPV headset.
- `MAVLINK_MODE_CUSTOM`: Stream nothing by default. Used when configuring streaming using MAVLink.
- `MAVLINK_MODE_MAGIC`: Same as `MAVLINK_MODE_CUSTOM`

> - `MAVLINK_MODE_CONFIG`: ?

Reviewer: This is used over USB by default, with a bit higher rates than NORMAL

Author: Thanks
```suggestion
- `MAVLINK_MODE_CONFIG`: Streaming over USB with higher rates than `MAVLINK_MODE_NORMAL`.
```

An example would be:

```sh
mavlink stream -u 14556 -s CA_TRAJECTORY -r 300
```

## Updating Ground Stations

Ultimately you'll want to use your new MAVLink interface by providing the corresponding ground station or MAVSDK implementation.

The important thing to remember here is that MAVLink requires that you use a version of the library that is built to the same definition (XML file).
So if you have created a custom message in PX4 you won't be able to use it unless you build QGC or MAVSDK with that same definition.

### Updating QGroundControl

You will need to [Build QGroundControl](https://docs.qgroundcontrol.com/master/en/qgc-dev-guide/getting_started/index.html) including a pre-built C library that contains your custom messages.

QGC uses a pre-built C library that must be located at [/qgroundcontrol/libs/mavlink/include/mavlink](https://github.com/mavlink/qgroundcontrol/tree/master/libs/mavlink/include/mavlink) in the QGC source.

By default this is pre-included as a submodule from <https://github.com/mavlink/c_library_v2> but you can [generate your own MAVLink Libraries](https://mavlink.io/en/getting_started/generate_libraries.html).

> QGC uses the ArduPilotMega.xml dialect by default, which includes **common.xml**.

Reviewer: ```suggestion
QGC uses the all.xml dialect by default, which includes **common.xml**.
```
